pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/MessagePriorityGrowableArrayBlockingQueue.java

```diff
@@ -41,64 +92,74 @@ public MessagePriorityGrowableArrayBlockingQueue(int initialCapacity) {
     queue =
         new PriorityBlockingQueue<>(
             initialCapacity,
-            new Comparator<Message>() {
+            new Comparator<MessageInfo>() {
               @Override
-              public int compare(Message o1, Message o2) {
-                int priority1 = getPriority(o1);
-                int priority2 = getPriority(o2);
+              public int compare(MessageInfo o1, MessageInfo o2) {
+                int priority1 = getPriority(o1.message);
+                int priority2 = getPriority(o2.message);
                 return Integer.compare(priority2, priority1);
               }
             });
   }
 
   @Override
   public Message remove() {
-    return queue.remove();
+    return toMessage(queue.remove());
+  }
+
+  private static Message toMessage(MessageInfo info) {
+    if (info == null) {
+      return null;
+    }
+    return info.message;
+  }
+  private static MessageInfo newInfo(Message msg) {
+    return new MessageInfo(msg, System.currentTimeMillis());
   }
 
   @Override
   public Message poll() {
-    return queue.poll();
+    return toMessage(queue.poll());
```

Comment:
do we really need to wrap ? this is adding lot of memory allocations and CPU processing time for all of the other operations

did you consider having a IdentityHashMap<Message, Long> instead of wrapping the Messages ?